lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp

```diff
@@ -0,0 +1,410 @@
+    BlockArgument inputValue =
+        moduleLike->getRegion(0).getArgument(inputProperty.index);
+    mapping.map(inputValue, parameterValue);
+  }
+
+  // Clone the property ops from the FIRRTL Class or Module to the OM Class.
+  SmallVector<Operation *> opsToErase;
+  OpBuilder builder = OpBuilder::atBlockBegin(classOp.getBodyBlock());
+  for (auto &op : moduleLike->getRegion(0).getOps()) {
+    // Check if any operand is a property.
+    auto propertyOperands = llvm::any_of(op.getOperandTypes(), [](Type type) {
+      return isa<PropertyType>(type);
+    });
+
+    // Check if any result is a property.
+    auto propertyResults = llvm::any_of(
+        op.getResultTypes(), [](Type type) { return isa<PropertyType>(type); });
+
+    // If there are no properties here, move along.
+    if (!propertyOperands && !propertyResults)
+      continue;
+
+    // Actually clone the op over to the OM Class.
+    builder.clone(op, mapping);
+
+    // In case this is a Module, remember to erase this op.
+    opsToErase.push_back(&op);
+  }
+
+  // Convert any output property assignments to Field ops.
+  for (auto op : llvm::make_early_inc_range(classOp.getOps<PropAssignOp>())) {
+    // Property assignments will currently be pointing back to the original
+    // FIRRTL Class for output ports.
+    auto outputPort = dyn_cast<BlockArgument>(op.getDest());
+    if (!outputPort)
+      continue;
+
+    // Get the original port name, create a Field, and erase the propassign.
+    auto name = moduleLike.getPortName(outputPort.getArgNumber());
+    builder.create<ClassFieldOp>(op.getLoc(), name, op.getSrc());
+    op.erase();
+  }
+
+  // If the module-like is a Class, completely erase it. Otherwise, erase just
+  // the property ports and ops.
+  if (isa<firrtl::ClassOp>(moduleLike)) {
+    moduleLike.erase();
+  } else {
+    // Erase ops in use before def order, thanks to FIRRTL's SSA regions.
+    for (auto *op : llvm::reverse(opsToErase))
+      op->erase();
+
+    // Erase property typed ports.
+    moduleLike.erasePorts(portsToErase);
+  }
+}
+
+// Update Object instantiations in a FIRRTL Module or OM Class.
+void LowerClassesPass::updateInstances(Operation *op) {
+  OpBuilder builder(op);
+  // For each Object instance.
+  for (auto firrtlObject : llvm::make_early_inc_range(
+           op->getRegion(0).getOps<firrtl::ObjectOp>())) {
+    // Collect its input actual parameters by finding any subfield ops that are
+    // assigned to. Take the source of the assignment as the actual parameter.
+    SmallVector<Value> actualParameters;
+    for (auto *user : firrtlObject->getUsers())
+      if (auto subfield = dyn_cast<ObjectSubfieldOp>(user))
+        for (auto *subfieldUser : subfield->getUsers())
+          if (auto propassign = dyn_cast<PropAssignOp>(subfieldUser))
+            if (propassign.getDest() == subfield.getResult())
+              actualParameters.push_back(propassign.getSrc());
```

Comment:
I couldn't actually write a test for this... but I think this is how it should work? Maybe I should remove it for now.

Reply:
I think this assumes that the ports are assigned in-order. You'll need to be able to map from old-port-index to new-argument-index, and then check what port was being assigned by looking at the subfield op's index.

Reply:
Yeah makes sense, I wasn't actually able to test this part anyway, so it's sort of a placeholder.